galsim/utilities.py

```diff
@@ -296,7 +296,134 @@ def _convertPositions(pos, units, func):
+
+    @returns a tuple of lists `(x_new, y_new)` with the thinned tabulation.
+    """
+    x = np.array(x)
+    f = np.array(f)
+
+    # Check for valid inputs
+    if len(x) != len(f):
+        raise ValueError("len(x) != len(f)")
+    if rel_err <= 0 or rel_err >= 1:
+        raise ValueError("rel_err must be between 0 and 1")
+    if not (np.diff(x) >= 0).all():
+        raise ValueError("input x is not sorted.")
+
+    # Check for trivial noop.
+    if len(x) <= 2:
+        # Nothing to do
+        return x,f
+
+    if trim_leading_zeros:
+        first = max(f.nonzero()[0][0]-1, 0)  # -1 to keep one non-redundant zero.
+        last = min(f.nonzero()[0][-1]+1, len(x)-1)  # +1 to keep one non-redundant zero.
+        x, f = x[first:last+1], f[first:last+1]
+
+    total_integ = np.trapz(abs(f), x)
+    if total_integ == 0:
+        return np.array([ x[0], x[-1] ]), np.array([ f[0], f[-1] ])
+    thresh = total_integ * rel_err
+
+    x_range = x[-1] - x[0]
+    if not preserve_range:
+        # Remove values from the front that integrate to less than thresh.
+        err_integ1 = 0.5 * (abs(f[0]) + abs(f[1])) * (x[1] - x[0])
+        k0 = 0
+        while k0 < len(x)-2 and err_integ1 < thresh * (x[k0+1]-x[0]) / x_range:
+            k0 = k0+1
+            err_integ1 += 0.5 * (abs(f[k0]) + abs(f[k0+1])) * (x[k0+1] - x[k0])
+        # Now the integral from 0 to k0+1 (inclusive) is a bit too large.
+        # That means k0 is the largest value we can use that will work as the starting value.
+
+        # Remove values from the back that integrate to less than thresh.
+        k1 = len(x)-1
+        err_integ2 = 0.5 * (abs(f[k1-1]) + abs(f[k1])) * (x[k1] - x[k1-1])
+        while k1 > k0 and err_integ2 < thresh * (x[-1]-x[k1-1]) / x_range:
+            k1 = k1-1
+            err_integ2 += 0.5 * (abs(f[k1-1]) + abs(f[k1])) * (x[k1] - x[k1-1])
+        # Now the integral from k1-1 to len(x)-1 (inclusive) is a bit too large.
+        # That means k1 is the smallest value we can use that will work as the ending value.
+
+        # Subtract the error so far from thresh
+        thresh -= np.trapz(abs(f[:k0]),x[:k0]) + np.trapz(abs(f[k1:]),x[k1:])
+
+        x = x[k0:k1+1]  # +1 since end of range is given as one-past-the-end.
+        f = f[k0:k1+1]
+
+        # And update x_range for the new values
+        x_range = x[-1] - x[0]
+
+    # Check again for noop after trimming endpoints.
+    if len(x) <= 2:
+        return x,f
+
+    # Thin interior points.  Start with no interior points and then greedily add them back in one at
+    # a time until relative error goal is met.
+    splitpoints = [0, len(x)-1]  # Start with single interval
+    m = (f[-1]-f[0])/(x[-1]-x[0]) # slope
+    fapprox = f[0]+m*(x-x[0])
+    errs = [np.trapz(np.abs(fapprox-f), x)]  # Current error for each interval
+    while sum(errs) > thresh:
+        # Find the worst current interval
+        index = np.argmax(errs)
```

Comment:
Another optimization note here, which I can do but might be a good learning exercise for you if you want to tackle it, is to change this to using a heap.  What you are doing here is called a Priority Queue.  And a heap is the standard efficient data structure for implementing this.  

In python, the heap functions are in `heapq`.  The data structure itself is a `list`, but you use the `heapq` commands to efficiently push and pop new values.  Here, you would `heappush` a tuple `(-err, index)`, which would use `-err` as the priority order.  Then `heappop` would get the item with the smallest `-err`, i.e. the largest `err`.